libp2p/autotls/acme/api.nim

```diff
@@ -0,0 +1,387 @@
+    let rawResponse =
+      await HttpClientRequestRef.get(session, acmeServerURL & "/directory").get().send()
+    let body = await rawResponse.getResponseBody()
+    body.to(ACMEDirectory)
+
+  ACMEApi(session: session, directory: directory, acmeServerURL: acmeServerURL)
+
+proc requestNonce(
+    self: ACMEApi
+): Future[Nonce] {.async: (raises: [ACMEError, CancelledError]).} =
+  try:
+    let resp =
+      await HttpClientRequestRef.get(self.session, self.directory.newNonce).get().send()
+    return resp.headers.getString("Replay-Nonce")
+  except HttpError as exc:
+    raise newException(ACMEError, "Failed to request new nonce from ACME server", exc)
+
+# TODO: save n and e in account so we don't have to recalculate every time
+proc acmeHeader(
+    self: ACMEApi, url: string, key: KeyPair, needsJwk: bool, kid: Opt[Kid]
+): Future[ACMERequestHeader] {.async: (raises: [ACMEError, CancelledError]).} =
+  if not needsJwk and kid.isNone:
+    raise newException(ACMEError, "kid not set")
+
+  let newNonce = await self.requestNonce()
+  if needsJwk:
+    # TODO: check if scheme is RSA
+    let pubkey = key.pubkey.rsakey
+    let nArray = @(getArray(pubkey.buffer, pubkey.key.n, pubkey.key.nlen))
+    let eArray = @(getArray(pubkey.buffer, pubkey.key.e, pubkey.key.elen))
+    ACMERequestHeader(
+      kind: ACMEJwkRequest,
+      alg: Alg,
+      typ: "JWT",
+      nonce: newNonce,
+      url: url,
+      jwk: JWK(kty: "RSA", n: base64UrlEncode(nArray), e: base64UrlEncode(eArray)),
+    )
+  else:
+    ACMERequestHeader(
+      kind: ACMEKidRequest,
+      alg: Alg,
+      typ: "JWT",
+      nonce: newNonce,
+      url: url,
+      kid: kid.get(),
+    )
+
+proc createSignedAcmeRequest(
+    self: ACMEApi,
+    url: string,
+    payload: auto,
+    key: KeyPair,
+    needsJwk: bool = false,
+    kid: Opt[Kid] = Opt.none(Kid),
+): Future[string] {.async: (raises: [ACMEError, CancelledError]).} =
+  if key.pubkey.scheme != PKScheme.RSA or key.seckey.scheme != PKScheme.RSA:
+    raise newException(ACMEError, "Unsupported signing key type")
+
+  let acmeHeader = await self.acmeHeader(url, key, needsJwk, kid)
+  handleError("createSignedAcmeRequest"):
+    var token = toJWT(%*{"header": acmeHeader, "claims": payload})
+    let derPrivKey = key.seckey.rsakey.getBytes.get
+    let pemPrivKey: string = pemEncode(derPrivKey, "PRIVATE KEY")
+    token.sign(pemPrivKey)
+    $token.toFlattenedJson()
+
+proc requestRegister*(
+    self: ACMEApi, key: KeyPair
+): Future[ACMERegisterResponse] {.async: (raises: [ACMEError, CancelledError]).} =
+  let registerRequest = ACMERegisterRequest(termsOfServiceAgreed: true)
+  handleError("acmeRegister"):
+    let payload = await self.createSignedAcmeRequest(
+      self.directory.newAccount, registerRequest, key, needsJwk = true
+    )
+    let rawResponse = await HttpClientRequestRef
+    .post(
+      self.session, self.directory.newAccount, body = payload, headers = ACMEHttpHeaders
+    )
+    .get()
+    .send()
```

Comment:
bonus:  consider having function for this these 6 lines become one `let resp = await self.post(url, payload)`